ql/src/java/org/apache/hadoop/hive/ql/exec/tez/TezSessionPoolManagerMetrics.java

```diff
@@ -0,0 +1,259 @@
+
+  private final TezSessionPoolManager poolManager;
+  private final Metrics metrics;
+  private ScheduledExecutorService mainExecutor;
+
+  // session pool metrics
+  @VisibleForTesting
+  final TezSessionMetric<Integer> runningTasksCount = new TezSessionMetric<>(TEZ_SESSION_METRIC_RUNNING_TASKS, 0);
+  @VisibleForTesting
+  final TezSessionMetric<Integer> pendingTasksCount = new TezSessionMetric<>(TEZ_SESSION_METRIC_PENDING_TASKS, 0);
+  @VisibleForTesting
+  final TezSessionMetric<Double> taskBacklogRatio = new TezSessionMetric<>(TEZ_SESSION_METRIC_TASK_BACKLOG_RATIO, 0.0);
+  private final TaskBacklogRatioMetric[] taskBacklogRatioDurations = Arrays.stream(TRACKED_TASK_BACKLOG_RATIOS)
+      .map(
+          ratio -> new TaskBacklogRatioMetric(String.format(TEZ_SESSION_METRIC_TASK_BACKLOG_RATIO_PRESENT_SINCE, ratio),
+              ratio))
+      .toArray(TaskBacklogRatioMetric[]::new);
+
+
+  public TezSessionPoolManagerMetrics(TezSessionPoolManager poolManager) {
+    this.poolManager = poolManager;
+    this.metrics = MetricsFactory.getInstance();
+  }
+
+  private static void waitForTasksToFinish(List<Future<?>> collectTasks, ExecutorService collectorExecutor) {
+    try {
+      CompletableFuture.allOf(collectTasks.toArray(new CompletableFuture[0])).get();
+    } catch (InterruptedException e) {
+      LOG.info("Interrupted while collecting session metrics", e);
+      Thread.currentThread().interrupt();
+    } catch (ExecutionException e) { // ExecutionException wraps the original exception
+      LOG.error("Exception while collecting session metrics", e.getCause());
+      throw new RuntimeException(e.getCause());
+    } finally {
+      collectorExecutor.shutdown();
+    }
+  }
+
+  public TezSessionPoolManagerMetrics start(HiveConf conf) {
+    if (metrics == null) {
+      LOG.warn(
+          "Metrics are not enabled, cannot start TezSessionPoolManagerMetrics (check {})",
+          HiveConf.ConfVars.HIVE_SERVER2_METRICS_ENABLED.varname);
+      return this;
+    }
+    this.mainExecutor = Executors.newSingleThreadScheduledExecutor(
+        new ThreadFactoryBuilder().setNameFormat("TezSessionPoolManagerMetrics worker thread").build());
+    int collectIntervalSeconds = HiveConf.getIntVar(conf,
+        HiveConf.ConfVars.HIVE_SERVER2_TEZ_SESSIONS_METRICS_COLLECTION_INTERVAL_SECONDS);
+    mainExecutor.scheduleAtFixedRate(this::collectMetrics, 0, collectIntervalSeconds, TimeUnit.SECONDS);
+    LOG.info("Starting TezSessionPoolManagerMetrics, collection interval: {}s", collectIntervalSeconds);
+
+    runningTasksCount.createGauge(metrics);
+    pendingTasksCount.createGauge(metrics);
+    taskBacklogRatio.createGauge(metrics);
+    for (TaskBacklogRatioMetric m : taskBacklogRatioDurations) {
+      m.createGauge(metrics);
+    }
+    clearMetrics(); //initial empty values for gauges
+    return this;
+  }
+
+  public void stop() {
+    if (this.mainExecutor != null) {
+      mainExecutor.shutdownNow();
```

Comment:
should this be ``shutdown`` or you want ``shutdownNow()`` only, moreover should we set `mainExecutor`=`null` or a check that it is already not shutdown or terminated. in case someone does stop twice or stop post shutdown

Reply:
hm, shutdown would be better, yes, as I don't need this call to attempt to stop actively executing tasks (what's shutdownNow is for), actually, I don't care about the running tasks here on a HS2 shutdown
regarding mainExecutor=null, it makes sense to simply make stop() idempotent